File: fenwick_tree/Cpp/fenwick_tree.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 100;

template <typename T>
struct fenwick_tree {
private:
   int n;
   vector <T> a;
public:
   void initialize(int k) {
      n=k;
      a.assign(n+1,0);
   }
   
   void update(int pos, T val) {
      for(;pos<=n;pos+=pos&(-pos)) a[pos]+=val;
   }
 
   T query(int pos) {
      T ans=0;
      for(;pos>=1;pos-=pos&(-pos)) ans+=a[pos];
      return ans;
   }
   
   T get(int from, int to) {
      return query(to)-query(from-1);
   }

   int lower_bound(T x){
      int index = 0;
      int sum = 0;
      for (int i = (log2(n) + 1); i >= 0; --i)
      {
         if((index+(1<<i)) <= n and (sum + a[index+(1<<i)]) < x){
            sum += a[index+(1<<i)];
            index += (1<<i);
         }
      }
      return index+1;
   }
};

fenwick_tree <int> bit;

int32_t main(){

   cout << "Initializing size to be 5" << endl;
   bit.initialize(5);

   cout << "Updating Value at 1" << endl;
   bit.update(1, 1);
   cout << "Updating Value at 2" << endl;
   bit.update(2, 2);
   cout << "Updating Value at 3" << endl;
   bit.update(3, 3);
   cout << "Updating Value at 4" << endl;
   bit.update(4, 4);
   cout << "Updating Value at 5" << endl;
   bit.update(5, 5);

   cout<<"The sum of values from 1 to 1 is : "<<endl;
   cout<<bit.query(1)<<endl;
   cout<<"The sum of values from 1 to 2 is : "<<endl;
   cout<<bit.query(2)<<endl;
   cout<<"The sum of values from 1 to 3 is : "<<endl;
   cout<<bit.query(3)<<endl;
   cout<<"The sum of values from 1 to 4 is : "<<endl;
   cout<<bit.query(4)<<endl;
   cout<<"The sum of values from 1 to 5 is : "<<endl;
   cout<<bit.query(5)<<endl;
}

File: fenwick_tree/Python/fenwick_tree.py
#!/usr/bin/env python3

"""
Binary Indexed Tree / Fenwick Tree
https://www.hackerearth.com/practice/notes/binary-indexed-tree-made-easy-2/
https://www.topcoder.com/community/data-science/data-science-tutorials/binary-indexed-trees/
https://www.youtube.com/watch?v=v_wj_mOAlig
https://www.youtube.com/watch?v=kPaJfAUwViY
"""


def update(index, value, array, bi_tree):
	"""
	Updates the binary indexed tree with the given value
	:param index: index at which the update is to be made
	:param value: the new element at the index
	:param array: the input array
	:param bi_tree: the array representation of the binary indexed tree
	:return: void
	"""

	while index < len(array):
		bi_tree[index] += value
		index += index & -index


def get_sum(index, bi_tree):
	"""
	Calculates the sum of the elements from the beginning to the index
	:param index: index till which the sum is to be calculated
	:param bi_tree: the array representation of the binary indexed tree
	:return: (integer) sum of the elements from beginning till index
	"""
	ans = 0

	while index > 0:
		ans += bi_tree[index]
		index -= index & -index

	return ans


def get_range_sum(left, right, bi_tree):
	"""
	Calculates the sum from the given range

	:param bi_tree: the array representation of the binary indexed tree
	:param left: left index of the range (1-indexed)
	:param right: right index of the range (1-indexed)
	:return: (integer) sum of the elements in the range
	"""
	ans = get_sum(right, bi_tree) - get_sum(left - 1, bi_tree)

	return ans


def main():
	n = int(input('Enter the number of elements: '))
	arr = [int(x) for x in input('Enter the {} elements of the array: '.format(n)).split()]
	arr.insert(0, 0)   # insert dummy node for 1-based indexing
	bit = [0 for i in range(n+1)]

	for index in range(1, n+1):
		update(index, arr[index], arr, bit)

	"""
	For range sum queries
		l, r = map(int, input('Enter the left and right indices for the range sum: ').split())
		print(get_range_sum(l, r, bit))

	For updating the binary indexed tree
		update(index, new_value - arr[index], arr, bit)
	"""


if __name__ == '__main__':
	main()
